internal/providers/msentraid/msentraid.go

```diff
@@ -255,3 +266,14 @@ func (c azureTokenCredential) GetToken(_ context.Context, _ policy.TokenRequestO
 		ExpiresOn: c.token.Expiry,
 	}, nil
 }
+
+func lowercaseASCII(s string) string {
+	var res []byte
+	for _, r := range []byte(s) {
+		if 'A' <= r && r <= 'Z' {
+			r += 'a' - 'A'
+		}
+		res = append(res, r)
+	}
+	return string(res)
+}
```

Comment:
> Would you mind pointing out a case where strings.ToLower(a) != strings.ToLower(b) would match something we don't want to match?

The "error when usernames differs in letters which fold to the same lowercase letter" test case is such a case

Reply:
Sure, I saw the test case. But is it something that can happen? If so, I'd rather we do two separate checks: one for valid ASCII usernames (which would fail if the string has special UTF-8 ones) and one for the case comparison. Even though `adduser` seems to allow adding users with UTF-8 characters, I'm pretty sure it would create issues with other system services and that would be very annoying to debug and fix.

Reply:
I'm also fine to fail with an error if the username contains non-ASCII characters. If Microsoft allows those names, we will notice at some point because users will complain, at which point we can revisit this. What do you think, @didrocks?

Reply:
> I didn’t find any general written rules though

Same

>  so maybe the only way would be testing different cases 

I'm not sure if it's worth spending time on this, especially since Microsoft Azure is such a complex beast that I don't think we can ever be sure that there's not some other way we didn't think of yet to create a user account or change its name. That's why I propose we enforce some rules ourselves which ensure that no two Entra users map to the same local user. We can decide if want to be more permissive (allow Unicode but don't fold non-ASCII characters) or more restrictive (allow only ASCII) in our rules.

>  (ofc, that one should work)

What are you referring to here with "that one"?

Reply:
We discussed this out-of-band and decided that we want to check that the usernames comply with the Microsoft Entra username policy.